nginx using boringssl start failed

when using boringSSL, it start failed. any suggestions?
```
    --with-openssl=../quiche/quiche/deps/boringssl \
    --with-quiche=../quiche \
```

nginx config
```
    init_by_lua_block {
        require("resty.acme.autossl").init({
            -- setting the following to true
            -- implies that you read and accepted https://letsencrypt.org/repository/
            tos_accepted = true,
            -- uncomment following for first time setup
            -- staging = true,
            -- uncomment following to enable RSA + ECC double cert
            domain_key_types = { "rsa", "ecc" },
            -- uncomment following to enable tls-alpn-01 challenge
            -- enabled_challenge_handlers = { "http-01", "tls-alpn-01" },
            account_key_path = "/etc/nginx/account.key",
            account_email = "n
@example.com",
            -- domain_whitelist = { "example.com" },
            domain_whitelist_callback = function(domain, is_new_cert_needed)
                return true
            end,
            renew_check_interval = 24 * 3600,
            storage_adapter = "file",
            storage_config = {
                dir = "/etc/nginx/acme",
            },
        })
    }

    init_worker_by_lua_block {
        require("resty.acme.autossl").init_worker()
    }
```

error log
```
nginx: [error] init_by_lua error: /etc/nginx/lualib/resty/openssl/version.lua:60: OpenSSL has encountered an error: /etc/nginx/lualib/resty/openssl/version.lua:45: /usr/local/lib/libluajit-5.1.so.2: undefined symbol: SSLeay; is OpenSSL library loaded?
stack traceback:
	[C]: in function 'error'
	/etc/nginx/lualib/resty/openssl/version.lua:60: in main chunk
	[C]: in function 'require'
	/etc/nginx/lualib/resty/acme/openssl.lua:4: in main chunk
	[C]: in function 'require'
	/etc/nginx/lualib/resty/acme/util.lua:4: in main chunk
	[C]: in function 'require'
	/etc/nginx/lualib/resty/acme/client.lua:3: in main chunk
	[C]: in function 'require'
	/etc/nginx/lualib/resty/acme/autossl.lua:2: in main chunk
	[C]: in function 'require'
	init_by_lua:2: in main chunk
```

nginx info
```
[root@8635137b486f lualib]# nginx -V
nginx version: nginx/1.20.2
built by gcc 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC)
built with OpenSSL 1.1.1 (compatible; BoringSSL) (running with BoringSSL)
TLS SNI support enabled
configure arguments: --with-cc-opt=-DTCP_FASTOPEN=23 --with-ld-opt='-ljemalloc -Wl,-rpath,/usr/local/lib' --prefix=/etc/nginx --conf-path=/etc/nginx/nginx.conf --sbin-path=/usr/sbin/nginx --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-log-path=/var/log/nginx/access.log --error-log-path=/var/log/nginx/error.log --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --with-file-aio --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_image_filter_module=dynamic --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-http_v3_module --with-http_xslt_module=dynamic --with-mail --with-mail_ssl_module --with-pcre-jit --with-stream --with-stream_ssl_module --with-threads --with-openssl=../quiche/quiche/deps/boringssl --with-quiche=../quiche --add-module=./module/ngx_brotli --add-module=./module/njs/nginx --add-module=./module/ngx_devel_kit-0.3.1 --add-module=./module/lua-nginx-module-0.10.20 --add-module=./module/stream-lua-nginx-module-0.0.10 --add-module=./module/nginx-rtmp-module --add-module=./module/nginx-client-module --add-module=./module/nginx-multiport-module --add-module=./module/nginx-toolkit-module --add-dynamic-module=./module/dynamic/ngx_waf --add-dynamic-module=./module/dynamic/ngx_http_geoip2_module --add-dynamic-module=./module/dynamic/echo-nginx-module --add-dynamic-module=./module/dynamic/headers-more-nginx-module --add-dynamic-module=./module/dynamic/srcache-nginx-module --add-dynamic-module=./module/dynamic/ngx_http_substitutions_filter_module
```

ldd /usr/sbin/nginx
```
[root@8635137b486f lualib]# ldd /usr/sbin/nginx
	linux-vdso.so.1 =>  (0x00007ffc9a114000)
	libjemalloc.so.2 => /usr/local/lib/libjemalloc.so.2 (0x00007fafe1e74000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fafe1c70000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fafe1a54000)
	libcrypt.so.1 => /lib64/libcrypt.so.1 (0x00007fafe181d000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fafe151b000)
	libluajit-5.1.so.2 => /usr/local/lib/libluajit-5.1.so.2 (0x00007fafe22af000)
	libpcre.so.1 => /usr/local/lib/libpcre.so.1 (0x00007fafe225b000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fafe1305000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fafe0f37000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fafe2119000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fafe0c2f000)
	libfreebl3.so => /lib64/libfreebl3.so (0x00007fafe0a2c000)
```

sign failed with error:
failed to create csr: failed to sign: x509.csr:sign: opt/lib-src/boringssl/crypto/evp/digestsign.c:107:error:06000077:public key routines:OPENSSL_internal:NO_DEFAULT_DIGEST

--- a/README.md
+++ b/README.md
@@ -54,6 +54,7 @@ Table of Contents
     + [pkey:set_parameters](#pkeyset_parameters)
     + [pkey:is_private](#pkeyis_private)
     + [pkey:get_key_type](#pkeyget_key_type)
+    + [pkey:get_default_digest_type](#pkeyget_default_digest_type)
     + [pkey:sign](#pkeysign)
     + [pkey:verify](#pkeyverify)
     + [pkey:encrypt](#pkeyencrypt)
@@ -942,6 +943,22 @@ local pkey, err = require("resty.openssl.pkey").new({type="X448"})
 
 ngx.say(require("cjson").encode(pkey:get_key_type()))
 -- outputs '{"ln":"X448","nid":1035,"sn":"X448","id":"1.3.101.111"}'
+```
+
+[Back to TOC](#table-of-contents)
+
+### pkey:get_default_digest_type
+
+**syntax**: *obj, err = pk:get_default_digest_type()*
+
+Returns a ASN1_OBJECT of key type of the private key as a table. An additional field `mandatory` is also
+returned in the table, if `mandatory` is true then other digests can not be used.
+
+```lua
+local pkey, err = require("resty.openssl.pkey").new()
+
+ngx.say(require("cjson").encode(pkey:get_default_digest_type()))
+-- outputs '{"ln":"sha256","nid":672,"id":"2.16.840.1.101.3.4.2.1","mandatory":false,"sn":"SHA256"}'
 ```
 
 [Back to TOC](#table-of-contents)

--- a/README.md
+++ b/README.md
@@ -2442,7 +2442,7 @@ Sign the certificate using the private key specified by `pkey`, which must be a
 parameter to set digest method, whichmust be a [resty.openssl.digest](#restyopenssldigest) instance.
 Returns a boolean indicating if signing is successful and error if any.
 
-In BoringSSL, `digest` must be set as no default digest is set.
+In BoringSSL when `digest` is not set it's fallback to `SHA256`.
 
 [Back to TOC](#table-of-contents)
 
@@ -2646,7 +2646,7 @@ Sign the certificate request using the private key specified by `pkey`, which mu
 parameter to set digest method, whichmust be a [resty.openssl.digest](#restyopenssldigest) instance.
 Returns a boolean indicating if signing is successful and error if any.
 
-In BoringSSL, `digest` must be set as no default digest is set.
+In BoringSSL when `digest` is not set it's fallback to `SHA256`.
 
 [Back to TOC](#table-of-contents)
 
@@ -2835,7 +2835,7 @@ Sign the CRL using the private key specified by `pkey`, which must be a
 parameter to set digest method, whichmust be a [resty.openssl.digest](#restyopenssldigest) instance.
 Returns a boolean indicating if signing is successful and error if any.
 
-In BoringSSL, `digest` must be set as no default digest is set.
+In BoringSSL when `digest` is not set it's fallback to `SHA256`.
 
 [Back to TOC](#table-of-contents)
 

--- a/examples/csr.lua
+++ b/examples/csr.lua
@@ -54,7 +54,7 @@ end
 -- create a EC key
 local pkey, err = require("resty.openssl.pkey").new({
   type = 'EC',
-  curve = 'prime192v1',
+  curve = 'prime256v1',
 })
 if err then
   error(err)

--- a/lib/resty/openssl/pkey.lua
+++ b/lib/resty/openssl/pkey.lua
@@ -33,6 +33,7 @@ local BORINGSSL = require("resty.openssl.version").BORINGSSL
 
 local ptr_of_uint = ctypes.ptr_of_uint
 local ptr_of_size_t = ctypes.ptr_of_size_t
+local ptr_of_int = ctypes.ptr_of_int
 
 local null = ctypes.null
 local load_pem_args = { null, null, null }
@@ -459,6 +460,24 @@ end
 
 function _M:get_key_type()
   return objects_lib.nid2table(self.key_type)
+end
+
+function _M:get_default_digest_type()
+  if BORINGSSL then
+    return nil, "BoringSSL doesn't have default digest for pkey"
+  end
+
+  local nid = ptr_of_int()
+  local code = C.EVP_PKEY_get_default_digest_nid(self.ctx, nid)
+  if code == -2 then
+    return nil, "operation is not supported by the public key algorithm"
+  elseif code <= 0 then
+    return nil, format_error("get_default_digest", code)
+  end
+
+  local ret = objects_lib.nid2table(nid[0])
+  ret.mandatory = code == 2
+  return ret
 end
 
 function _M:get_provider_name()

--- a/lib/resty/openssl/x509/crl.lua
+++ b/lib/resty/openssl/x509/crl.lua
@@ -16,6 +16,7 @@ local format_error = require("resty.openssl.err").format_error
 local version = require("resty.openssl.version")
 local OPENSSL_10 = version.OPENSSL_10
 local OPENSSL_11_OR_LATER = version.OPENSSL_11_OR_LATER
+local BORINGSSL = version.BORINGSSL
 local BORINGSSL_110 = version.BORINGSSL_110 -- used in boringssl-fips-20190808
 
 local accessors = {}
@@ -194,16 +195,20 @@ function _M:sign(pkey, digest)
     return false, "x509.crl:sign: expect a pkey instance at #1"
   end
 
+  local digest_algo
   if digest then
     if not digest_lib.istype(digest) then
       return false, "x509.crl:sign: expect a digest instance at #2"
     elseif not digest.algo then
       return false, "x509.crl:sign: expect a digest instance to have algo member"
     end
+    digest_algo = digest.algo
+  elseif BORINGSSL then
+    digest_algo = C.EVP_get_digestbyname('sha256')
   end
 
   -- returns size of signature if success
-  if C.X509_CRL_sign(self.ctx, pkey.ctx, digest and digest.algo) == 0 then
+  if C.X509_CRL_sign(self.ctx, pkey.ctx, digest_algo) == 0 then
     return false, format_error("x509.crl:sign")
   end
 

--- a/lib/resty/openssl/x509/csr.lua
+++ b/lib/resty/openssl/x509/csr.lua
@@ -20,6 +20,7 @@ local format_error = require("resty.openssl.err").format_error
 local version = require("resty.openssl.version")
 local OPENSSL_10 = version.OPENSSL_10
 local OPENSSL_11_OR_LATER = version.OPENSSL_11_OR_LATER
+local BORINGSSL = version.BORINGSSL
 local BORINGSSL_110 = version.BORINGSSL_110 -- used in boringssl-fips-20190808
 
 local accessors = {}
@@ -323,16 +324,20 @@ function _M:sign(pkey, digest)
     return false, "x509.csr:sign: expect a pkey instance at #1"
   end
 
+  local digest_algo
   if digest then
     if not digest_lib.istype(digest) then
       return false, "x509.csr:sign: expect a digest instance at #2"
     elseif not digest.algo then
       return false, "x509.csr:sign: expect a digest instance to have algo member"
     end
+    digest_algo = digest.algo
+  elseif BORINGSSL then
+    digest_algo = C.EVP_get_digestbyname('sha256')
   end
 
   -- returns size of signature if success
-  if C.X509_REQ_sign(self.ctx, pkey.ctx, digest and digest.algo) == 0 then
+  if C.X509_REQ_sign(self.ctx, pkey.ctx, digest_algo) == 0 then
     return false, format_error("x509.csr:sign")
   end
 

--- a/lib/resty/openssl/x509/init.lua
+++ b/lib/resty/openssl/x509/init.lua
@@ -23,6 +23,7 @@ local version = require("resty.openssl.version")
 local OPENSSL_10 = version.OPENSSL_10
 local OPENSSL_11_OR_LATER = version.OPENSSL_11_OR_LATER
 local OPENSSL_30 = version.OPENSSL_30
+local BORINGSSL = version.BORINGSSL
 local BORINGSSL_110 = version.BORINGSSL_110 -- used in boringssl-fips-20190808
 
 -- accessors provides an openssl version neutral interface to lua layer
@@ -380,16 +381,20 @@ function _M:sign(pkey, digest)
     return false, "x509:sign: expect a pkey instance at #1"
   end
 
+  local digest_algo
   if digest then
     if not digest_lib.istype(digest) then
       return false, "x509:sign: expect a digest instance at #2"
     elseif not digest.algo then
       return false, "x509:sign: expect a digest instance to have algo member"
     end
+    digest_algo = digest.algo
+  elseif BORINGSSL then
+    digest_algo = C.EVP_get_digestbyname('sha256')
   end
 
   -- returns size of signature if success
-  if C.X509_sign(self.ctx, pkey.ctx, digest and digest.algo) == 0 then
+  if C.X509_sign(self.ctx, pkey.ctx, digest_algo) == 0 then
     return false, format_error("x509:sign")
   end
 

--- a/scripts/templates/x509_functions.j2
+++ b/scripts/templates/x509_functions.j2
@@ -5,16 +5,20 @@ function _M:sign(pkey, digest)
     return false, "{{modname}}:sign: expect a pkey instance at #1"
   end
 
+  local digest_algo
   if digest then
     if not digest_lib.istype(digest) then
       return false, "{{modname}}:sign: expect a digest instance at #2"
-    elseif not digest.dtyp then
-      return false, "{{modname}}:sign: expect a digest instance to have dtyp member"
+    elseif not digest.algo then
+      return false, "{{modname}}:sign: expect a digest instance to have algo member"
     end
+    digest_algo = digest.algo
+  elseif BORINGSSL then
+    digest_algo = C.EVP_get_digestbyname('sha256')
   end
 
   -- returns size of signature if success
-  if C.{{ module.type }}_sign(self.ctx, pkey.ctx, digest and digest.dtyp) == 0 then
+  if C.{{ module.type }}_sign(self.ctx, pkey.ctx, digest_algo) == 0 then
     return false, format_error("{{ modname }}:sign")
   end
 

--- a/t/openssl/pkey.t
+++ b/t/openssl/pkey.t
@@ -1202,3 +1202,26 @@ default
 "
 --- no_error_log
 [error]
+
+=== TEST 37: Get default digest type
+--- http_config eval: $::HttpConfig
+--- config
+    location =/t {
+        content_by_lua_block {
+            if require("resty.openssl.version").BORINGSSL then
+                ngx.say('BROKEN sha256 BROKEN')
+                ngx.exit(0)
+            end
+
+            local pkey = require("resty.openssl.pkey")
+            local p = myassert(pkey.new({ type = "EC" }))
+            local algo = myassert(p:get_default_digest_type())
+            ngx.say(require("cjson").encode(algo))
+        }
+    }
+--- request
+    GET /t
+--- response_body_like
+.+sha256.+
+--- no_error_log
+[error]

